Cs/sc 1789 update user mapping

Charl1996 opened the pull request:
Relates to [this](https://github.com/dimagi/commcare-hq/pull/31596) PR.

This PR implements what was suggested [here](https://github.com/dimagi/commcare-hq/pull/31596#discussion_r869562937). In this PR the new field (assigned_location_ids) is simply added to the `user_mapping.py` file. The change will be applied by the included migration as well as the field population.

Regarding backwards compatibility:
An additional filter has been added to the existing function, `mobile_user_ids_at_locations`, which ensures that mobile users will be returned. However, there are also some places throughout the code where the `.location` filter is applied (9 places). For all instances it seems the `.location` filter was applied with the assumption that web users will be returned so the `.mobile_users` filter (or whichever other filter) was applied separately where necessary, except in the following two instances which I'm unsure of:
-  [Project Health](https://github.com/dimagi/commcare-hq/blob/d58c94383e9883ba67c7e6355e9cd213f14617a5/corehq/apps/reports/standard/project_health.py#L291) (not sure what the intent is for the `previous_months_summary` method)
- [Report filters controllers](https://github.com/dimagi/commcare-hq/blob/c4db48c894ea3a09a99cefda99d3e9c155973be4/corehq/apps/reports/filters/controllers.py#L112) (although looking at [this](https://github.com/dimagi/commcare-hq/blob/c4db48c894ea3a09a99cefda99d3e9c155973be4/corehq/apps/reports/filters/controllers.py#L114) method it seems the intent is to include _all_ users)


Copy of related PR's details below:

## Product Description
No visible changes

## Technical Summary
This PR addresses [this](https://github.com/dimagi/commcare-hq/blob/02ef417fa33e0dcc9c81bb1b1c997c1d41d44bb6/corehq/apps/es/users.py#L170) comment by adding the `assigned_location_ids` to user_mapping.py, which is necessary to make the result of the UserES().location filter more accurate, i.e. to include web users and not just the mobile users at that location. This is important for tasks such as [update_users_at_locations](https://github.com/dimagi/commcare-hq/blob/b80d7781672d399962375fba689173b68438ade4/corehq/apps/locations/tasks.py#L97) (which is executed when a location is removed and updates the users assigned to that location accordingly).

Because [mobile_user_ids_at_locations(...)](https://github.com/dimagi/commcare-hq/blob/4ca04a02a7c61e6d3a4970f119a537ce571a94b2/corehq/apps/locations/dbaccessors.py#L93) relies on this "bug" that existed (and is called from a bunch of places), I'm ensuring that the UserES().location filter still returns the expected results by simply allowing an argument to the location filter function in order to include the web users in the filter results.   
 
## Feature Flag
Not FF specific

## Safety Assurance

### Safety story
Unit test + local testing

### Automated test coverage
Unit test added which specifically shows that web users are also included in the location filter result.

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
- [ ] The migrations in this code can be safely applied first independently of the code

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Charl1996 committed:
Add assigned_location_ids to user_mapping and update using migration

Charl1996 committed:
Ensure function returns mobile users

Charl1996 commented:
Just reiterating what I stated in the description...

I'm unsure what the intent is for the following two instances (i.e. to consider _all_ users, or only mobile users), so I'm not sure if I should add some additional logic to only filter the intended users in conjunction with the `location` filter:
- [Project Health](https://github.com/dimagi/commcare-hq/blob/d58c94383e9883ba67c7e6355e9cd213f14617a5/corehq/apps/reports/standard/project_health.py#L291) (not sure what the intent is for the previous_months_summary method)
- [Report filters controllers](https://github.com/dimagi/commcare-hq/blob/c4db48c894ea3a09a99cefda99d3e9c155973be4/corehq/apps/reports/filters/controllers.py#L112) (although looking at [this](https://github.com/dimagi/commcare-hq/blob/c4db48c894ea3a09a99cefda99d3e9c155973be4/corehq/apps/reports/filters/controllers.py#L114) method it seems the intent is to include all users)

I'm hoping somebody with more context on these two instances can provide some insights.

Charl1996 committed:
Lock migration

snopoke reviewed (commented):
Code changes look fine but I'm a bit concerned about performing the reindex as part of the migration. Is there an estimate for how long this will take on production?

What is the impact of the reindex not happening? As it stands the mapping update could go through and the reindex could fail for some reason.

joxl commented on corehq/ex-submodules/pillowtop/migrations/0007_update_user_mapping_with_assigned_location_ids.py:
```diff
@@ -0,0 +1,28 @@
+    return get_all_docs_with_doc_types(db=CouchUser.get_db(), doc_types=["WebUser"])
+
+
+def reindex_web_users(*args, **kwargs):
```
This operation isn't needed on a fresh install.

```suggestion
@skip_on_fresh_install
def reindex_web_users(*args, **kwargs):
```

joxl commented on corehq/ex-submodules/pillowtop/migrations/0007_update_user_mapping_with_assigned_location_ids.py:
```diff
@@ -0,0 +1,28 @@
+from django.db import migrations
+
+from corehq.pillows.mappings.user_mapping import USER_INDEX_INFO
+from corehq.util.django_migrations import update_es_mapping
```
Decorator needed for suggestion below.
```suggestion
from corehq.util.django_migrations import skip_on_fresh_install, update_es_mapping
```

joxl commented on corehq/ex-submodules/pillowtop/migrations/0007_update_user_mapping_with_assigned_location_ids.py:
```diff
@@ -0,0 +1,28 @@
+
+    operations = [
+        update_es_mapping(USER_INDEX_INFO.index),
+        migrations.RunPython(reindex_web_users),
```
This operation does not have reversal logic.
```suggestion
        migrations.RunPython(
            reindex_web_users,
            reverse_code=migrations.RunPython.noop,
        ),
```

joxl reviewed (commented):
A few tweaks to the migration.

calellowitz commented:
I want to repeat my request from the last PR to break this into two different pull requests, to be rolled out at different times. I think it is both safer and easier to separate the mapping change from the migration. That will also allow us to run the migration at any time, once the mapping change is out, and ensure consistent data at all times. Even if all goes perfectly with this PR, there is still a period of time between when the migration is started and when the code switches over where these queries will return incorrect data.

Instead I think the order should be
1) Create a PR with the mapping change and the changes to existing queries to ensure we can handle partially reindexed web users, but no data migration
2) Create a second PR with the data migration to reindex all web users, and any new functionality that relies on the presence of web users

We can run the migration from step 2 at any point in a private release so it does not block deploy as it is unlikely to be quick.

esoergel commented on corehq/ex-submodules/pillowtop/migrations/0007_update_user_mapping_with_assigned_location_ids.py:
```diff
@@ -0,0 +1,28 @@
+
+def reindex_web_users(*args, **kwargs):
+    for user in all_webusers():
+        CouchUser.wrap(user).save()
```
Also there is a lot of stuff that happens on user save - sync to SQL, maybe fire repeaters, sync usercase, update in ES (synchronously), and some others. `update_user_in_es` should be much faster.

Charl1996 committed:
Add management command to sync web users in ES

Charl1996 commented on corehq/apps/locations/tasks.py:
```diff
@@ -99,7 +99,7 @@ def update_users_at_locations(domain, location_ids, supply_point_ids, ancestor_i
     """
     from corehq.apps.users.models import CouchUser, update_fixture_status_for_users
-    from corehq.apps.locations.dbaccessors import mobile_user_ids_at_locations
+    from corehq.apps.locations.dbaccessors import user_ids_at_locations
```
At the moment this will return the same result as `mobile_user_ids_at_locations`. After the data migration this will also include the actual web users.

Charl1996 replied:
Actually, if the mapping update is made and new web users are created, this `user_ids_at_locations` function will also include those web users, which is OK.

Charl1996 committed:
Fix test import

Charl1996 commented:
Changes made to this PR:
1) Removed the migration
2) Added management command which simply syncs ES with Couch

Future plans:
After this PR is merged, I'll add another PR to add the migration (maybe in private release) which updates the mapping and invokes the management command to sync the we users' data.
